README.md
# php-appboy

A PHP client for sending push notifications via the [Appboy](https://www.appboy.com/documentation/REST_API) API

[![Author](http://img.shields.io/badge/author-@superbalist-blue.svg?style=flat-square)](https://twitter.com/superbalist)
[![Build Status](https://img.shields.io/travis/Superbalist/php-appboy/master.svg?style=flat-square)](https://travis-ci.org/Superbalist/php-appboy)
[![StyleCI](https://styleci.io/repos/92749823/shield?branch=master)](https://styleci.io/repos/92749823)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Packagist Version](https://img.shields.io/packagist/v/superbalist/php-appboy.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-appboy)
[![Total Downloads](https://img.shields.io/packagist/dt/superbalist/php-appboy.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-appboy)


## Installation

```bash
composer require superbalist/php-appboy
```

## Usage

```php
use GuzzleHttp\Client;
use Superbalist\Appboy\Appboy;
use Superbalist\Appboy\NotificationBuilder;
use Superbalist\Appboy\ScheduledNotificationBuilder;
use Superbalist\Appboy\Messages\AndroidMessageBuilder;
use Superbalist\Appboy\Messages\AppleMessageBuilder;

$client = new Client();
$appboy = new Appboy($client, 'your-app-group-id');

// send a push message
$appboy->sendMessage(
    (new NotificationBuilder())
        ->toUsers([1, 2])
        ->setCampaign('my_campaign')
        ->ignoreFrequencyCapping()
        ->setSubscriptionState('opted_in')
        ->withMessages([
            'apple_push' => (new AppleMessageBuilder())
                ->setAlert('Hello World!')
                ->setSound('custom_sound')
                ->withExtraAttributes(['is_test' => true])
                ->setCategory('shipping_notification')
                ->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
                ->setUri('http://superbalist.com')
                ->setMessageVariation('group_a')
                ->setAsset('file://image.jpg', 'jpg')
                ->build(),
            'android_push' => (new AndroidMessageBuilder())
                ->setAlert('Hello World!')
                ->setTitle('Message Title')
                ->withExtraAttributes(['is_test' => true])
                ->setMessageVariation('group_a')
                ->setPriority(2)
                ->setCollapseKey('shipment_1234')
                ->setSound('custom_sound')
                ->setUri('http://superbalist.com')
                ->setSummaryText('This is a summary line')
                ->setTimeToLive(60)
                ->setNotificationId(18456)
                ->setPushIconImageUrl('http://link/to/asset.jpg')
                ->setAccentColour(16777215)
                ->build(),
        ])
        ->build()
);

// schedule a push message
$appboy->scheduleMessage(
    (new ScheduledNotificationBuilder())
        ->toUsers([1, 2])
        ->setCampaign('my_campaign')
        ->ignoreFrequencyCapping()
        ->setSubscriptionState('opted_in')
        ->withMessage(
            'apple_push',
            (new AppleMessageBuilder())
                ->setAlert('Hello World!')
                ->setSound('custom_sound')
                ->withExtraAttributes(['is_test' => true])
                ->setCategory('shipping_notification')
                ->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
                ->setUri('http://superbalist.com')
                ->setMessageVariation('group_a')
                ->setAsset('file://image.jpg', 'jpg')
                ->build()
        )
        ->sendsAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
        ->build()
```


changelog.md
# Changelog

## 1.0.1 - 2017-06-29

* Fix 400 bad request when empty 'extra' params are passed

## 1.0.0 - 2017-05-30

* Initial release

tests/AppboyTest.php
<?php

namespace Tests;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Mockery;
use PHPUnit\Framework\TestCase;
use Superbalist\Appboy\Appboy;

class AppboyTest extends TestCase
{
    public function testSetGetUri()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = new Appboy($guzzleClient, 'ABC123');
        $this->assertEquals('https://api.appboy.com', $appboy->getUri());
        $appboy->setUri('http://127.0.0.1');
        $this->assertEquals('http://127.0.0.1', $appboy->getUri());

        $appboy = new Appboy($guzzleClient, 'ABC123', 'http://192.168.0.1');
        $this->assertEquals('http://192.168.0.1', $appboy->getUri());
    }

    public function testSetGetAppGroupId()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = new Appboy($guzzleClient, 'ABC123');
        $this->assertEquals('ABC123', $appboy->getAppGroupId());
        $appboy->setAppGroupId('9998');
        $this->assertEquals('9998', $appboy->getAppGroupId());
    }

    public function testGetClient()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = new Appboy($guzzleClient, 'ABC123');
        $this->assertSame($guzzleClient, $appboy->getClient());
    }

    public function testMakeBaseUri()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = new Appboy($guzzleClient, 'ABC123');
        $this->assertEquals('https://api.appboy.com/messages/send', $appboy->makeBaseUri('/messages/send'));
        $this->assertEquals('https://api.appboy.com/messages/send', $appboy->makeBaseUri('messages/send'));
        $this->assertEquals('https://api.appboy.com/messages/send', $appboy->makeBaseUri('messages/send/'));
    }

    public function testPost()
    {
        $request = new Request(
            'POST',
            'https://api.appboy.com/messages/send',
            [
                'Content-Type' => 'application/json',
            ],
            json_encode(['lorem' => 'ipsum'])
        );

        $guzzleClient = Mockery::mock(Client::class);
        $appboy = Mockery::mock('\Superbalist\Appboy\Appboy[createRequest,sendRequest]', [$guzzleClient, 'ABC123']);
        $appboy->shouldAllowMockingProtectedMethods();
        $appboy->shouldReceive('createRequest')
            ->withArgs([
                'POST',
                'messages/send',
                json_encode(['lorem' => 'ipsum']),
                ['Content-Type' => 'application/json'],
            ])
            ->once()
            ->andReturn($request);
        $appboy->shouldReceive('sendRequest')
            ->with($request)
            ->once();

        $appboy->post('messages/send', ['lorem' => 'ipsum']);
    }

    public function testSendMessage()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = Mockery::mock('\Superbalist\Appboy\Appboy[post]', [$guzzleClient, 'ABC123']);
        $appboy->shouldReceive('post')
            ->withArgs([
                'messages/send',
                [
                    'app_group_id' => 'ABC123',
                    'blah' => 'bleh',
                ],
            ])
            ->once();
        $appboy->sendMessage(['blah' => 'bleh']);
    }

    public function testScheduleMessage()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $appboy = Mockery::mock('\Superbalist\Appboy\Appboy[post]', [$guzzleClient, 'ABC123']);
        $appboy->shouldReceive('post')
            ->withArgs([
                'messages/schedule/create',
                [
                    'app_group_id' => 'ABC123',
                    'blah' => 'bleh',
                ],
            ])
            ->once();
        $appboy->scheduleMessage(['blah' => 'bleh']);
    }
}


tests/AppleMessageBuilderTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use Superbalist\Appboy\Messages\AppleMessageBuilder;

class AppleMessageBuilderTest extends TestCase
{
    public function testSetBadgeCount()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setBadgeCount(3);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('badge', $params);
        $this->assertEquals(3, $params['badge']);
    }

    public function testSetAlert()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setAlert('Hello World!');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('alert', $params);
        $this->assertEquals('Hello World!', $params['alert']);
    }

    public function testSetSound()
    {
        $builder = new AppleMessageBuilder();
        $params = $builder->build();
        $this->assertArrayHasKey('sound', $params);
        $this->assertEquals('default', $params['sound']);

        $return = $builder->setSound('custom_sound');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('sound', $params);
        $this->assertEquals('custom_sound', $params['sound']);
    }

    public function testWithExtraAttributes()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->withExtraAttributes(['is_test' => true, 'show_banner' => false]);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('extra', $params);
        $this->assertEquals(['is_test' => true, 'show_banner' => false], $params['extra']);

        $builder->withExtraAttributes(['version' => 'abc']);
        $params = $builder->build();
        $this->assertArrayHasKey('extra', $params);
        $this->assertEquals(['version' => 'abc'], $params['extra']);

        $builder->withExtraAttributes([]);
        $params = $builder->build();
        $this->assertArrayNotHasKey('extra', $params);
    }

    public function testAddExtraAttribute()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->addExtraAttribute('is_test', true);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('extra', $params);
        $this->assertEquals(['is_test' => true], $params['extra']);

        $builder->withExtraAttributes(['show_banner' => false]);
        $params = $builder->build();
        $this->assertEquals(['show_banner' => false], $params['extra']);

        $builder->addExtraAttribute('is_test', true);
        $builder->addExtraAttribute('version', 'abc');
        $params = $builder->build();
        $this->assertEquals(
            [
                'show_banner' => false,
                'is_test' => true,
                'version' => 'abc',
            ],
            $params['extra']
        );
    }

    public function testAddExtraAttributes()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->addExtraAttributes(['is_test' => true, 'show_banner' => false]);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('extra', $params);
        $this->assertEquals(['is_test' => true, 'show_banner' => false], $params['extra']);

        $builder->withExtraAttributes(['version' => 'abc']);
        $params = $builder->build();
        $this->assertEquals(['version' => 'abc'], $params['extra']);

        $builder->addExtraAttributes(['is_test' => true]);
        $builder->addExtraAttributes(['show_banner' => false]);
        $params = $builder->build();
        $this->assertEquals(
            [
                'version' => 'abc',
                'is_test' => true,
                'show_banner' => false,
            ],
            $params['extra']
        );
    }

    public function testSetCategory()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setCategory('shipping_notification');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('category', $params);
        $this->assertEquals('shipping_notification', $params['category']);
    }

    public function testExpiresAt()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')));
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('expiry', $params);
        $this->assertEquals('2017-05-29T10:00:00+02:00', $params['expiry']);
    }

    public function testSetUri()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setUri('http://superbalist.com');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('custom_uri', $params);
        $this->assertEquals('http://superbalist.com', $params['custom_uri']);
    }

    public function testSetMessageVariation()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setMessageVariation('group_a');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('message_variation_id', $params);
        $this->assertEquals('group_a', $params['message_variation_id']);
    }

    public function testSetAsset()
    {
        $builder = new AppleMessageBuilder();
        $return = $builder->setAsset('file://image.jpg', 'jpg');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('asset_url', $params);
        $this->assertArrayHasKey('asset_file_type', $params);
        $this->assertEquals('file://image.jpg', $params['asset_url']);
        $this->assertEquals('jpg', $params['asset_file_type']);
    }

    public function testBuild()
    {
        $builder = new AppleMessageBuilder();
        $params = $builder->setBadgeCount(3)
            ->setAlert('Hello World!')
            ->setSound('custom_sound')
            ->withExtraAttributes(['is_test' => true])
            ->setCategory('shipping_notification')
            ->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
            ->setUri('http://superbalist.com')
            ->setMessageVariation('group_a')
            ->setAsset('file://image.jpg', 'jpg')
            ->build();

        $expected = [
            'badge' => 3,
            'alert' => 'Hello World!',
            'sound' => 'custom_sound',
            'extra' => [
                'is_test' => true,
            ],
            'category' => 'shipping_notification',
            'expiry' => '2017-05-29T10:00:00+02:00',
            'custom_uri' => 'http://superbalist.com',
            'message_variation_id' => 'group_a',
            'asset_url' => 'file://image.jpg',
            'asset_file_type' => 'jpg',
        ];
        $this->assertEquals($expected, $params);
    }
}


tests/NotificationBuilderTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use Superbalist\Appboy\NotificationBuilder;

class NotificationBuilderTest extends TestCase
{
    public function testToUsers()
    {
        $builder = new NotificationBuilder();
        $return = $builder->toUsers([1, 2]);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('external_user_ids', $params);
        $this->assertEquals([1, 2], $params['external_user_ids']);

        $builder->toUsers(2);
        $params = $builder->build();
        $this->assertArrayHasKey('external_user_ids', $params);
        $this->assertEquals([2], $params['external_user_ids']);
    }

    public function testToUser()
    {
        $builder = new NotificationBuilder();
        $return = $builder->toUser(2);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('external_user_ids', $params);
        $this->assertEquals([2], $params['external_user_ids']);
    }

    public function testToSegment()
    {
        $builder = new NotificationBuilder();
        $return = $builder->toSegment('abc123');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('segment_id', $params);
        $this->assertEquals('abc123', $params['segment_id']);
    }

    public function testSetCampaign()
    {
        $builder = new NotificationBuilder();
        $return = $builder->setCampaign('my_campaign');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('campaign_id', $params);
        $this->assertEquals('my_campaign', $params['campaign_id']);
    }

    public function testIgnoreFrequencyCapping()
    {
        $builder = new NotificationBuilder();
        $params = $builder->build();
        $this->assertArrayHasKey('override_frequency_capping', $params);
        $this->assertFalse($params['override_frequency_capping']);

        $return = $builder->ignoreFrequencyCapping();
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('override_frequency_capping', $params);
        $this->assertTrue($params['override_frequency_capping']);
    }

    public function testRespectFrequencyCapping()
    {
        $builder = new NotificationBuilder();
        $builder->ignoreFrequencyCapping();
        $return = $builder->respectFrequencyCapping();
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('override_frequency_capping', $params);
        $this->assertFalse($params['override_frequency_capping']);
    }

    public function testSetSubscriptionState()
    {
        $builder = new NotificationBuilder();
        $params = $builder->build();
        $this->assertArrayHasKey('recipient_subscription_state', $params);
        $this->assertEquals('subscribed', $params['recipient_subscription_state']);

        $return = $builder->setSubscriptionState('all');
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('recipient_subscription_state', $params);
        $this->assertEquals('all', $params['recipient_subscription_state']);
    }

    public function testWithMessages()
    {
        $builder = new NotificationBuilder();
        $return = $builder->withMessages([
            'apple_push' => ['alert' => 'Message 1'],
            'android_push' => ['alert' => 'Message 2'],
        ]);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('messages', $params);
        $expected = [
            'apple_push' => ['alert' => 'Message 1'],
            'android_push' => ['alert' => 'Message 2'],
        ];
        $this->assertEquals($expected, $params['messages']);

        $builder->withMessages(['apple_push' => ['alert' => 'Message 3']]);
        $params = $builder->build();
        $this->assertEquals(['apple_push' => ['alert' => 'Message 3']], $params['messages']);
    }

    public function testWithMessage()
    {
        $builder = new NotificationBuilder();
        $return = $builder->withMessage('apple_push', ['alert' => 'Message 1']);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('messages', $params);
        $this->assertEquals(['apple_push' => ['alert' => 'Message 1']], $params['messages']);

        $builder->withMessage('android_push', ['alert' => 'Message 2']);
        $params = $builder->build();
        $this->assertEquals(['android_push' => ['alert' => 'Message 2']], $params['messages']);
    }

    public function testAddMessage()
    {
        $builder = new NotificationBuilder();
        $return = $builder->addMessage('apple_push', ['alert' => 'Message 1']);
        $this->assertSame($builder, $return);
        $params = $builder->build();
        $this->assertArrayHasKey('messages', $params);
        $this->assertEquals(['apple_push' => ['alert' => 'Message 1']], $params['messages']);

        $builder->withMessage('android_push', ['alert' => 'Message 2']);
        $params = $builder->build();
        $this->assertEquals(['android_push' => ['alert' => 'Message 2']], $params['messages']);

        $builder->addMessage('apple_push', ['alert' => 'Message 3']);
        $builder->addMessage('kindle_push', ['alert' => 'Message 4']);
        $params = $builder->build();
        $this->assertEquals(
            [
                'android_push' => ['alert' => 'Message 2'],
                'apple_push' => ['alert' => 'Message 3'],
                'kindle_push' => ['alert' => 'Message 4'],
            ],
            $params['messages']
        );
    }

    public function testBuild()
    {
        $builder = new NotificationBuilder();
        $params = $builder->toUsers([1, 2])
            ->setCampaign('my_campaign')
            ->ignoreFrequencyCapping()
            ->setSubscriptionState('opted_in')
            ->withMessage('apple_push', ['alert' => 'Hello World!'])
            ->build();

        $expected = [
            'external_user_ids' => [1, 2],
            'campaign_id' => 'my_campaign',
            'override_frequency_capping' => true,
            'recipient_subscription_state' => 'opted_in',
            'messages' => [
                'apple_push' => ['alert' => 'Hello World!'],
            ],
        ];
        $this->assertEquals($expected, $params);
    }
}


examples/SendPush.php
<?php

// usage:
// APPBOY_APP_GROUP_ID='your app group id' php examples/SendPush.php

include __DIR__ . '/../vendor/autoload.php';

use GuzzleHttp\Client;
use Superbalist\Appboy\Appboy;
use Superbalist\Appboy\Messages\AndroidMessageBuilder;
use Superbalist\Appboy\Messages\AppleMessageBuilder;
use Superbalist\Appboy\NotificationBuilder;

$client = new Client();
$appboy = new Appboy($client, getenv('APPBOY_APP_GROUP_ID'));

// send a push message
$response = $appboy->sendMessage(
    (new NotificationBuilder())
        ->toUser(2)
        ->withMessages([
            'apple_push' => (new AppleMessageBuilder())
                ->setAlert('This is a test message')
                ->withExtraAttributes(['is_test' => true])
                ->setCategory('matthew_test')
                ->expiresAt(new \DateTime('+1 hour', new \DateTimeZone('Africa/Johannesburg')))
                ->setUri('http://superbalist.com')
                ->build(),
            'android_push' => (new AndroidMessageBuilder())
                ->setAlert('This is a test message')
                ->setTitle('Message Title')
                ->withExtraAttributes(['is_test' => true])
                ->setUri('http://superbalist.com')
                ->setSummaryText('This is a summary line')
                ->build(),
        ])
        ->build()
);
var_dump($response);


src/NotificationBuilder.php
<?php

namespace Superbalist\Appboy;

class NotificationBuilder
{
    /**
     * @var array
     */
    protected $externalUserIds = [];

    /**
     * @var string
     */
    protected $segmentId;

    /**
     * @var string
     */
    protected $campaignId;

    /**
     * @var bool
     */
    protected $overrideFrequencyCapping = false;

    /**
     * @var string
     */
    protected $recipientSubscriptionState = 'subscribed';

    /**
     * @var array
     */
    protected $messages = [];

    /**
     * Create a new NotificationBuilder.
     *
     * **Example**
     *
     * ```php
     * $builder = new NotificationBuilder();
     * $params = $builder->toUsers([1, 2])
     *     ->setCampaign('my_campaign')
     *     ->ignoreFrequencyCapping()
     *     ->setSubscriptionState('opted_in')
     *     ->withMessage(['alert' => 'Hello World!'])
     *     ->build();
     * ```
     */
    public function __construct()
    {
        //
    }

    /**
     * Set the external user ids to send to.
     *
     * @param array|mixed $ids
     *
     * @return NotificationBuilder
     *
     * @see https://www.appboy.com/documentation/REST_API/#external-user-id
     */
    public function toUsers($ids)
    {
        if (!is_array($ids)) {
            $ids = [$ids];
        }

        $this->externalUserIds = $ids;
        return $this;
    }

    /**
     * Set the external user id to send to.
     *
     * @param mixed $id
     *
     * @return NotificationBuilder
     *
     * @see https://www.appboy.com/documentation/REST_API/#external-user-id
     */
    public function toUser($id)
    {
        return $this->toUsers($id);
    }

    /**
     * Set the segment to send to.
     *
     * @param string $segmentId
     *
     * @return NotificationBuilder
     *
     * @see https://www.appboy.com/documentation/REST_API/#segment-identifier
     */
    public function toSegment($segmentId)
    {
        $this->segmentId = $segmentId;
        return $this;
    }

    /**
     * Set the campaign identifier.
     *
     * @param string $campaignId
     *
     * @return NotificationBuilder
     *
     * @see https://www.appboy.com/documentation/REST_API/#campaign-identifier
     */
    public function setCampaign($campaignId)
    {
        $this->campaignId = $campaignId;
        return $this;
    }

    /**
     * Ignore the frequency capping for the campaign send.
     *
     * This will set the override_frequency_capping param to true, default is false.
     *
     * @return NotificationBuilder
     */
    public function ignoreFrequencyCapping()
    {
        $this->overrideFrequencyCapping = true;
        return $this;
    }

    /**
     * Respect the frequency capping for the campaign send.
     *
     * This will set the override_frequency_capping param to false, which is the default.
     *
     * @return NotificationBuilder
     */
    public function respectFrequencyCapping()
    {
        $this->overrideFrequencyCapping = false;
        return $this;
    }

    /**
     * Set the required subscription state of the audience.
     *
     * The state can be one of:
     * 1. opted_in - only users who have opted in
     * 2. subscribed - only users who have subscribed or are opted in
     * 3. all - all users, including unsubscribed users
     *
     * The default state is 'subscribed' only.
     *
     * @param string $state (opted_int|subscribed|all)
     *
     * @return NotificationBuilder
     */
    public function setSubscriptionState($state)
    {
        $this->recipientSubscriptionState = $state;
        return $this;
    }

    /**
     * Set the message push objects.
     *
     * @param array $messages
     *
     * @return NotificationBuilder
     */
    public function withMessages(array $messages)
    {
        $this->messages = $messages;
        return $this;
    }

    /**
     * Set the message object.
     *
     * @param string $type (apple_push|android_push|...)
     * @param array $message
     *
     * @return NotificationBuilder
     */
    public function withMessage($type, array $message)
    {
        $this->messages = [
            $type => $message,
        ];
        return $this;
    }

    /**
     * Add a message object.
     *
     * @param string $type (apple_push|android_push|...)
     * @param array $message
     *
     * @return NotificationBuilder
     */
    public function addMessage($type, array $message)
    {
        $this->messages[$type] = $message;
        return $this;
    }

    /**
     * Build the notification payload.
     *
     * @return array
     */
    public function build()
    {
        $params = [
            'external_user_ids' => $this->externalUserIds,
            'segment_id' => $this->segmentId,
            'campaign_id' => $this->campaignId,
            'override_frequency_capping' => $this->overrideFrequencyCapping,
            'recipient_subscription_state' => $this->recipientSubscriptionState,
            'messages' => $this->messages,
        ];

        return array_filter($params, function ($param) {
            return $param !== null;
        });
    }
}


src/ScheduledNotificationBuilder.php
<?php

namespace Superbalist\Appboy;

use DateTime;

class ScheduledNotificationBuilder extends NotificationBuilder
{
    /**
     * @var DateTime
     */
    protected $sendDate;

    /**
     * Create a new ScheduledNotificationBuilder.
     *
     * **Example**
     *
     * ```php
     * $builder = new ScheduledNotificationBuilder();
     * $params = $builder->toUsers([1, 2])
     *     ->setCampaign('my_campaign')
     *     ->ignoreFrequencyCapping()
     *     ->setSubscriptionState('opted_in')
     *     ->withMessage(['alert' => 'Hello World!'])
     *     ->sendsAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
     *     ->build();
     * ```
     */
    public function __construct()
    {
        //
    }

    /**
     * Set the send date/time.
     *
     * @param DateTime $date
     *
     * @return ScheduledNotificationBuilder
     */
    public function sendsAt(DateTime $date)
    {
        $this->sendDate = $date;
        return $this;
    }

    /**
     * Build the notification payload.
     *
     * @return array
     */
    public function build()
    {
        $params = parent::build();

        if ($this->sendDate) {
            $params['schedule'] = [
                'time' => $this->sendDate->format('c'),
            ];
        }

        return $params;
    }
}


src/Appboy.php
<?php

namespace Superbalist\Appboy;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Psr\Http\Message\RequestInterface;

class Appboy
{
    /**
     * @var Client
     */
    protected $client;

    /**
     * @var string
     */
    protected $uri;

    /**
     * @var string
     */
    protected $appGroupId;

    /**
     * Create a new Appboy client.
     *
     * **Example**
     *
     * ```php
     * $client = new \GuzzleHttp\Client();
     * $appboy = new Appboy($client, 'ABC123');
     * ```
     *
     * @param Client $client
     * @param string $appGroupId
     * @param string $uri
     */
    public function __construct(Client $client, $appGroupId, $uri = 'https://api.appboy.com')
    {
        $this->client = $client;
        $this->appGroupId = $appGroupId;
        $this->uri = $uri;
    }

    /**
     * Set the uri.
     *
     * @param string $uri
     */
    public function setUri($uri)
    {
        $this->uri = $uri;
    }

    /**
     * Return the uri.
     *
     * @return string
     */
    public function getUri()
    {
        return $this->uri;
    }

    /**
     * Set the App Group Identifier.
     *
     * The App Group Identifier can be found in the developer console settings.
     *
     * @param string $appGroupId
     *
     * @see https://www.appboy.com/documentation/REST_API/#app-group-identifier-explanation
     */
    public function setAppGroupId($appGroupId)
    {
        $this->appGroupId = $appGroupId;
    }

    /**
     * Return the App Group Identifier.
     *
     * @return string
     */
    public function getAppGroupId()
    {
        return $this->appGroupId;
    }

    /**
     * @return Client
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * Return the full uri to an API end-point.
     *
     * @param string $endpoint
     *
     * @return string
     */
    public function makeBaseUri($endpoint)
    {
        return rtrim($this->uri, '/') . '/' . trim($endpoint, '/');
    }

    /**
     * Create an HTTP request object.
     *
     * @param string $method
     * @param string $endpoint
     * @param mixed $body
     * @param array $headers
     *
     * @return Request
     */
    protected function createRequest($method, $endpoint, $body = null, array $headers = [])
    {
        $uri = $this->makeBaseUri($endpoint);
        return new Request($method, $uri, $headers, $body);
    }

    /**
     * Make an HTTP POST request.
     *
     * @param string $endpoint
     * @param array $data
     *
     * @return array
     */
    public function post($endpoint, array $data = [])
    {
        $body = json_encode($data);
        $request = $this->createRequest('POST', $endpoint, $body, ['Content-Type' => 'application/json']);
        return $this->sendRequest($request);
    }

    /**
     * Send an HTTP request.
     *
     * @param RequestInterface $request
     *
     * @throws \Exception
     *
     * @return array
     */
    protected function sendRequest(RequestInterface $request)
    {
        $response = $this->client->send($request);
        return json_decode($response->getBody(), true);
    }

    /**
     * Send a message.
     *
     * **Example**
     *
     * ```php
     * $appboy->sendMessage(
     *     (new NotificationBuilder())
     *         ->toUsers([1, 2])
     *         ->setCampaign('my_campaign')
     *         ->ignoreFrequencyCapping()
     *         ->setSubscriptionState('opted_in')
     *         ->withMessage(
     *             'apple_push',
     *             (new AppleMessageBuilder())
     *                 ->setAlert('Hello World!')
     *                 ->setSound('custom_sound')
     *                 ->withExtraAttributes(['is_test' => true])
     *                 ->setCategory('shipping_notification')
     *                 ->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
     *                 ->setUri('http://superbalist.com')
     *                 ->setMessageVariation('group_a')
     *                 ->setAsset('file://image.jpg', 'jpg')
     *                 ->build()
     *         )
     *         ->build()
     * );
     * ```
     *
     * @param array $params
     *
     * @return array
     */
    public function sendMessage(array $params)
    {
        $params['app_group_id'] = $this->appGroupId;
        return $this->post('messages/send', $params);
    }

    /**
     * Schedule a message.
     *
     * **Example**
     *
     * ```php
     * $appboy->scheduleMessage(
     *     (new ScheduledNotificationBuilder())
     *         ->toUsers([1, 2])
     *         ->setCampaign('my_campaign')
     *         ->ignoreFrequencyCapping()
     *         ->setSubscriptionState('opted_in')
     *         ->withMessage(
     *             'apple_push',
     *             (new AppleMessageBuilder())
     *                 ->setAlert('Hello World!')
     *                 ->setSound('custom_sound')
     *                 ->withExtraAttributes(['is_test' => true])
     *                 ->setCategory('shipping_notification')
     *                 ->expiresAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
     *                 ->setUri('http://superbalist.com')
     *                 ->setMessageVariation('group_a')
     *                 ->setAsset('file://image.jpg', 'jpg')
     *                 ->build()
     *         )
     *         ->sendsAt(new \DateTime('2017-05-29 10:00:00', new \DateTimeZone('Africa/Johannesburg')))
     *         ->build()
     * );
     * ```
     *
     * @param array $params
     *
     * @return array
     */
    public function scheduleMessage(array $params)
    {
        $params['app_group_id'] = $this->appGroupId;
        return $this->post('messages/schedule/create', $params);
    }
}
